handler.go

```diff
@@ -0,0 +1,339 @@
+)
+
+// streamRequestSupplierFn is a function that returns the stream message body
+// if supplier has no more messages it must return io.EOF
+type streamRequestSupplierFn func() (requestBody []byte, err error)
+
+// streamResponseHandlerFn is a function that receives the stream response body
+type streamResponseHandlerFn func(responseBody []byte) error
+
+type requestHandler struct {
+	out        output
+	logger     *zap.Logger
+	opts       Options
+	transport  transport.Transport
+	resolved   resolvedProtocolEncoding
+	serializer encoding.Serializer
+	body       io.Reader
+	headers    map[string]string
+}
+
+func (r requestHandler) handle() {
+	if isStreamingMethod(r.serializer) {
+		r.handleStreamRequest()
+		return
+	}
+
+	r.handleUnaryRequest()
+}
+
+// handleUnaryRequest launches initial unary request and unary benchmark
+func (r requestHandler) handleUnaryRequest() {
+	reqInput, err := ioutil.ReadAll(r.body)
+	if err != nil {
+		r.out.Fatalf("Failed while reading the input: %v\n", err)
+	}
+
+	req, err := r.serializer.Request(reqInput)
+	if err != nil {
+		r.out.Fatalf("Failed while serializing the input: %v\n", err)
+	}
+
+	req, err = prepareRequest(req, r.headers, r.opts)
+	if err != nil {
+		r.out.Fatalf("Failed while preparing the request: %v\n", err)
+	}
+
+	if r.shouldMakeInitialRequest() {
+		makeInitialRequest(r.out, r.transport, r.serializer, req)
+	}
+
+	runBenchmark(r.out, r.logger, r.opts, r.resolved, benchmarkMethod{
+		serializer: r.serializer,
+		req:        req,
+	})
+}
+
+// handleStreamRequest launches initial stream request and stream benchmark
+func (r requestHandler) handleStreamRequest() {
+	streamSerializer, ok := r.serializer.(encoding.StreamSerializer)
+	if !ok {
+		r.out.Fatalf("Serializer does not support streaming: %v\n", r.serializer.Encoding())
+	}
+
+	streamReq, streamMsgReader, err := streamSerializer.StreamRequest(r.body)
+	if err != nil {
+		r.out.Fatalf("Failed to create streaming request: %v\n", err)
+	}
+
+	streamReq.Request, err = prepareRequest(streamReq.Request, r.headers, r.opts)
+	if err != nil {
+		r.out.Fatalf("Failed while preparing the request: %v\n", err)
+	}
+
+	nextBodyFn := func() ([]byte, error) {
+		msg, err := streamMsgReader.NextBody()
+		if err == io.EOF {
+			return nil, err
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Failed while reading stream input: %v", err)
+		}
+
+		return msg, nil
+	}
+
+	responseHandlerFn := func(body []byte) error {
+		res, err := r.serializer.Response(&transport.Response{Body: body})
+		if err != nil {
+			return fmt.Errorf("Failed while serializing stream response: %v", err)
+		}
+
+		bs, err := json.MarshalIndent(res, "", "  ")
+		if err != nil {
+			return fmt.Errorf("Failed to convert map to JSON: %v\nMap: %+v", err, res)
+		}
+
+		r.out.Printf("%s\n\n", bs)
+		return nil
+	}
+
+	if r.shouldMakeInitialRequest() {
+		if err = makeStreamRequest(r.transport, streamReq, streamSerializer, nextBodyFn, responseHandlerFn); err != nil {
```

Comment:
Why did you choose to have `func makeStreamRequest` over `func(r requestHandler) makeStreamRequest`? (you would need to pass less argument to the functions and `responseHandlerFn` and `nextBodyFn` could be attached to the receiver. 

Thoughts?

(I guess you did this because it is like this for the unary request, but this does not mean we should follow the same pattern here - or we can change it)

Reply:
Kept `makeStreamRequest` separately as it is invoked in the benchmark (next PR), handler struct could be sparsely constructed but that would be confusing since we wouldn't know what all fields this method would use
https://github.com/yarpc/yab/blob/c1572df9f116de9812e4b0cd13c9c5f9fa740c53/bench_method.go#L105

Reply:
I see, we could refactor this I believe. We could give `m` to `benchmarkMethod` and `m` could implement a specific interface. 

Let's discuss offline about this if we want to do it in another diff